### PR TITLE
Opaygo step by step librairies and functions integration

### PR DESCRIPTION
**Pull request by Peguy-WANDA**
Add in main.cpp the function GetDataFromToken(...) from opaygo_decoder.h

### CODE REVIEW

File: Firmware code/smart_energy_meter/lib/opaygo_decoder/extended/opaygo_core_extended.c [hunk 1]
```diff
@@ -28,7 +28,7 @@ uint64_t ConvertHashToTokenExtended(uint64_t this_hash) {
     return result;
 }
 
-uint64_t GenerateOPAYGOTokenExtended(uint64_t LastToken, char SECRET_KEY[16]) {
+uint64_t GenerateOPAYGOTokenExtended(uint64_t LastToken, unsigned char SECRET_KEY[16]) {
```
Comment (frimic-solarly) [inline Firmware code/smart_energy_meter/lib/opaygo_decoder/extended/opaygo_core_extended.c:31]: @dmohns : This is a test comment.

**Comment by Peguy-WANDA** [comment]
Hi @dmohns, i would like to know how to organize the functions and libraries of Opaygo that we are going to use? Two possibilities are available to us:
1-Create in the src folder, new files containing the functions and libraries necessary for opaygo;
2-Add the opaygo functions and libraries in the existing openSmartMeter files in the src folder.

Personally I suggest that we opt for the first possibility to have a more structured and more organized code. But for now I'm waiting for your answer to organize the rest of the code. THANKS

**Comment by dmohns** [comment]
> Hi @dmohns, i would like to know how to organize the functions and libraries of Opaygo that we are going to use? Two possibilities are available to us: 1-Create in the src folder, new files containing the functions and libraries necessary for opaygo; 2-Add the opaygo functions and libraries in the existing openSmartMeter files in the src folder.
> 
> Personally I suggest that we opt for the first possibility to have a more structured and more organized code. But for now I'm waiting for your answer to organize the rest of the code. THANKS

Hey @Peguy-WANDA sorry for the late reply 🙈 

Unfortunately, I'm not sure I understand the question. Let me clarify:

The `opaygo-decoder` should be treated as a **library**. What I mean by this is: We should not modify any files from the library, but use it the way it is right now. That's why it's currently sitting in the `lib/` folder of the project.

Moving forward, the code will entirely disappear from the OpenSmartMeter repository and will be pulled in via dependencies of `platform.ini`. However, this is only possible after: https://github.com/EnAccess/OpenPAYGO-HW/issues/14 has been solved.

In case you are talking about "wrapper" functions that are needed in OpenSmartMeter firmware to utilize PAYGO functions: These function can basically go wherever you see them fit best. I, personally, would probably put it to a similar place where the current token functions sit, i.e. `token_management.hpp`. But not a strong opinion here.

File: Firmware code/smart_energy_meter/src/token_management.hpp [hunk 2]
```diff
@@ -6,7 +6,7 @@
 // Arduino base libraries
 
 // third party libraries
-#include <Keypad.h>
+#include "../.pio/libdeps/bluepill_f103c8_128k/Keypad/src/Keypad.h"
```
Comment (dmohns) [inline Firmware code/smart_energy_meter/src/token_management.hpp:9]: This shouldn't be here 🤔 Looks like something is not working correctly on your setup, can you double check?

**Comment by dmohns** [comment]
Another comment. It looks like your branched out from a very old version of the `main` branch.

image
 
(Blue in the screenshot)

Could you rebase or update your branch to the latest version of `main`?

File: Firmware code/smart_energy_meter/src/token_management.hpp (hunk 2 above)
Comment (Peguy-WANDA) [reply Firmware code/smart_energy_meter/src/token_management.hpp:9]: OK thanks for your answer

**Review by dmohns — COMMENTED** [review]
Few comments

File: Firmware code/smart_energy_meter/.clang-format [hunk 3]
```diff
@@ -10,3 +10,4 @@ DerivePointerAlignment: false
 # Always break after if to get accurate coverage
 AllowShortIfStatementsOnASingleLine: false
 AllowShortLoopsOnASingleLine: false
+
```
Comment (dmohns) [inline Firmware code/smart_energy_meter/.clang-format:13]: Can you revert that change so the file is untouched?

File: Firmware code/smart_energy_meter/src/credit.hpp [hunk 4]
```diff
@@ -1,5 +1,6 @@
-#pragma once
-
+//#pragma once
+#ifndef credit_hpp
+#define credit_hpp
```
Comment (dmohns) [inline Firmware code/smart_energy_meter/src/credit.hpp:3]: Why is this needed?

**Review by dmohns — COMMENTED** [review]
Hey, few comments on the PR, would be great if you could address them

1. Could you revert all whitespace changes to files that are unrelated to this PR? If you think any of those changes would be better for code readability please feel free to send a separate PR. 
2. I see you are changing all include guards from `#pragma once` to `#ifndef` any specific reason why you need this? As above, please leave these untouched in this PR, but open another one, if you think this will make the code better.

Other than that, looks good to me!! Thanks for all the work

File: Firmware code/smart_energy_meter/src/token_management.hpp [hunk 5]
```diff
@@ -89,12 +91,14 @@ void STS_keypad() {
     dt = 0;
     sts_value = 0;
     delay(20);
-    lcd.clear();
     buss();
+    lcd.clear();
     lcd.setCursor(0, 0);
```
Comment (dmohns) [inline Firmware code/smart_energy_meter/src/token_management.hpp:96]: I don't know what the `buss()` function does. Could you explain why it is needed for `lcd.clear()` to be called after `buss()` instead of before?

Comment (Peguy-WANDA) [reply Firmware code/smart_energy_meter/src/token_management.hpp]: Ok I will add comments in the code to explain it

File: Firmware code/smart_energy_meter/src/credit.hpp (hunk 4 above)
Comment (Peguy-WANDA) [reply Firmware code/smart_energy_meter/src/credit.hpp]: The two syntaxe are doing the same things but there is a difference between them.
The #ifndef/#define/#endif trick works on any C compiler, and on some of them it speeds up the compilation process. The #pragma trick is non-standards, and only works on few C compilers, and may result in different semantics in those that do not support it.
When I use #pragma trick, I have many warning, but with  #ifndef/#define/#endif trick the warnings disappear

Comment (Peguy-WANDA) [reply Firmware code/smart_energy_meter/src/credit.hpp]: but for this PR I'm just going to put back #pragma once as you wanted

File: Firmware code/smart_energy_meter/src/token_management.hpp (hunk 5 above)
Comment (Peguy-WANDA) [reply Firmware code/smart_energy_meter/src/token_management.hpp]: The buss() function is there just to emit a sound to signal to the user that he can already enter the token in sts mode

**Comment by Peguy-WANDA** [comment]
Daniel I think as of now everything is good for this pull request, thank you for your comments, feel free to leave more if needed.
Have a nice weekend

File: firmware/src/time_management.hpp [hunk 6]
```diff
@@ -12,6 +12,10 @@
 #include "thingsboard.hpp"
 
 RTC_DS1307 rtc;
+#define ACTIVE_UNTIL_NVRAM_ADDRESS \
+  0  // 56 bytes address from 0 to 55 in the NVRAM
+#define TIME_INITIALIZATION_NVRAM_ADDRESS 4
```
Comment (dmohns) [inline firmware/src/time_management.hpp:17]: Where are these used? If they aren't, can we remove them?

File: firmware/src/opaygo_functions.hpp [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+#endif
+
+// Device parameters location in Flash/EEPROM
+unsigned int TokenCount_eeprom_location = 6;
+unsigned int UsedTokens_eeprom_location = 8;
+unsigned int PAYGEnabled_eeprom_location = 10;
+unsigned int ActiveUntil_eeprom_location = 11;
+unsigned int TokenEntryLockedUntil_eeprom_location = 15;
+unsigned int LAST_TIME_STAMP_ADDRESS = 27;
+unsigned int NB_DISCONNECTIONS_ADDRESS = 32;
```
Comment (dmohns) [inline firmware/src/opaygo_functions.hpp:41]: What do you think about making some like a "global" EEPROM map? We don't have to do this now in this PR, but as an idea?

For example, in `global_defined` we could include all the EEPROM locations as macros?

File: firmware/src/opaygo_functions.hpp [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+   *Returns the unixtime in seconds of Now
+   */
+  DateTime now = rtc.now();
+  uint32_t nowInSeconds =
+      (now.unixtime() -
+       timeInitializationRtc);  // we substract the init time so that it's
+                                // easier to read when debugging, and works the
+                                // same as Arduino Time mgt
+  return (nowInSeconds);
+}
```
Comment (dmohns) [inline firmware/src/opaygo_functions.hpp:132]: This isn't PAYGO related, but time related, should it go in `time_management.hpp` instead?

File: firmware/src/opaygo_functions.hpp [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+
+void SetTime(int ActivationValue) {
+  uint32_t Now = GetTimeInSeconds();
+  int NumberOfSecondsToActivate = (ActivationValue * 3600 * 24) / TIME_DIVIDER;
+
+  ActiveUntil =
+      Now +
+      NumberOfSecondsToActivate;  // We set the number of days (converted in
+                                  // seconds for to compare to our RTC time)
+}
```
Comment (dmohns) [inline firmware/src/opaygo_functions.hpp:202]: This isn't PAYGO related, but time related, should it go in `time_management.hpp` instead?

File: firmware/src/opaygo_functions.hpp [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+    storeUint32InNvram(timeInitializationRtc,
+                       TIME_INITIALIZATION_NVRAM_ADDRESS);
+  } else {  // a disconnection happened
+    timeInitializationRtc = nvramCheck;
+    incrementNbDisconnectionsEeprom();
+    storeTimeStampEEPROM(
+        GetTimeInSeconds());  // to avoid counting twice the disconnection if
+                              // the previous TimeStamp was too long ago
+  }
+}
```
Comment (dmohns) [inline firmware/src/opaygo_functions.hpp:313]: This isn't PAYGO related, but time related, should it go in `time_management.hpp` instead?

File: firmware/src/sts_token.hpp [hunk 11]
```diff
@@ -58,8 +58,8 @@ void STStoken_decode() {
   String identifier = sts_third.substring(6, 7);
   String sts_encode = sts_data.substring(15, 20);
   long tariff_MT_NO = sts_encode.toInt();
-  long tariff_gotten = ((tariff_MT_NO) - (meter_no * multiplier));
-  long trueMT_NO = tariff_MT_NO - tariff;
+  // long tariff_gotten = ((tariff_MT_NO) - (meter_no * multiplier));
```
Comment (dmohns) [inline firmware/src/sts_token.hpp:61]: If you comment out a line, please add a comment saying why this is commented out.

File: firmware/src/opaygo_functions.hpp [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+
+#define BLINK_PERIOD 250
+#define DEBUG
+
+#define STAR_KEY -1
+#define HASH_KEY -2
+#define NON_ACCEPTED_KEY -3
+
+#define TIME_DIVIDER 1
+// #define RESTRICTED_DIGIT_SET_MODE
```
Comment (dmohns) [inline firmware/src/opaygo_functions.hpp:26]: Can this comment be removed?
